### DISCUSSION TITLE
Reuse data fetcher without having to declare it on every concrete type

### DISCUSSION THREAD

**Discussion by testower** [graphql-java/graphql-java#3209] [Q&A] [answered]
Before graphql-java v12 it was possible to reuse a data fetcher for a field definition that was common across types in the following way:

```
// Define a common field definition
        GraphQLFieldDefinition idFieldDefinition = newFieldDefinition()
                .name("id")
                .type(new GraphQLNonNull(GraphQLID))
                .dataFetcher(env -> ... unified way to resolve this field ...)
                .build();
                
// Define a common "abstract type"

        GraphQLObjectType identifiedEntityObjectType = newObject().name("IdentifiedEntity")
                .field(idFieldDefinition)
                .build();
                
// Reuse this type for a "concrete" type

        GraphQLObjectType fooObjectType = newObject(identifiedEntityObjectType).name("Foo")
                ...
                .build();
```

As of version 12, this no longer works, since data fetchers are registered in a map that is keyed by the name of the object types, and this information is seemingly lost during the "newObject" factory invocation.

I have tried some different approaches, like converting the IdentifiedEntity into an interface, and registering a data fetcher on the interface type via the code registry instead, but this doesn't work either:

```
        GraphQLInterfaceType identifiedEntityInterface = new GraphQLInterfaceType.Builder()
                .name("IdentifiedEntity")
                .field(idFieldDefinition)
                .build();
                
                
        GraphQLObjectType fooObjectType = newObject()
               .withInterface(identifiedEntityInterface)
               .name("Foo")
               .field(idFieldDefinition)
                ...
                .build();
                

         GraphQLCodeRegistry.newCodeRegistry()
                .dataFetcher(
                        identifiedEntityInterface,
                        idFieldDefinition,
                        ... add common data fetcher logic here ...
                )
                .... snipped ...
                .build();
```

This doesn't work either. The data fetcher registered via code registry is never invoked, and I get null for the id field.

The only thing that seems to work is to register the data fetcher explicitly on each "concrete" type. When I have a dozen types that needs this, it feels cumbersome. Is there any other way?

**Comment by bbakerman** [comment, marked as answer]
If we had out time again - we would not have allowed the runtime GraphqlFieldDefinitions to have data fetchers on them.

This mixes two concerns - the representation of the type system that is the schema and the code behind a field that fetches values.

We introduced `graphql.schema.GraphQLCodeRegistry` in 2018 as the place where "data fetchers" a registered against a type and a field.  See also `graphql.schema.FieldCoordinates`

So you can re-use the code of the data fetcher but you must indeed register it per Object type and field name.

data fetchers ONLY ever apply to an object type and its fields.  There is no data fetcher on an interface.  Interfaces (and unions) get resolved down to an Object type at runtime and then its fields are fetched via looking from the object type and field name to a data fetcher in the code registry

We still allow `graphql.schema.GraphQLFieldDefinition.Builder#dataFetcher` because we didnt want to break the world but during schema creation we traverse the schema and lift the data fetchers from here and place them into a code registry.

I am suprised this does not work for you but either way setting data fetchers onto field definitions was deprecated on the `2018-12-03` and will not be coming back so I think you should embrace a code registry pattern

**Reply by testower** [reply]
By all means, it's not that I don't want to embrace the code registry pattern, I just wanted to check if there was a way to avoid having to register the data fetcher per Object type, in case I had missed something that should have been very obvious too me.

**Reply by testower** [reply]
And yes I realize I'm very late to the party. I am in the process of upgrading an application that was stuck on v11. But better late than never I guess...
